### PR TITLE
chore(ssa refactor): Add code to handle array equality

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

## Problem\*

Resolves 

## Summary\*

This PR sets out to

### Example

Before:

```

```

After:

```

```

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs [hunk 1]
```diff
@@ -178,6 +184,15 @@ impl BrilligGen {
...
             }
+            Instruction::ArrayGet { array, index } => {
+                let array_ptr = self.convert_ssa_value(*array, dfg);
+                let index = self.convert_ssa_value(*index, dfg);
+
+                let result_ids = dfg.instruction_results(instruction_id);
+                let destination = self.get_or_create_register(result_ids[0]);
+
+                self.context.array_get(array_ptr, index, destination)
+            }
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_gen.rs:195]: Will remove this is not needed for array_eq example

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs [hunk 2]
```diff
@@ -187,20 +202,75 @@ impl BrilligGen {
...
+            element_type.len(),
+            1,
+            "array has a composite type as its element, which we do not support yet"
+        );
+        let atomic_type = element_type[0].clone();
+
+        match operator {
+            BinaryOp::Eq => {
+                // TODO:
+                // Add a for loop which does roughly
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_gen.rs:259]: Currently this code returns 1 for array equality;  just need to implement the for-loop noted below

File: crates/noirc_evaluator/src/brillig/brillig_ir.rs [hunk 3]
```diff
@@ -88,6 +123,18 @@ impl BrilligContext {
...
+        // Note: We could insert a const instruction to initialize the register
+        // because the VM will not expand the register space automatically.
+        //
+        // In most cases, the register created is used in another instruction
+        // which will cause the VM to expand the register space, but this is not
+        // a guarantee.
+        //
+        // TODO: check if the above can be true if we just return a constant for example
+        // TODO from a program
+        // self.const_instruction(register, 0u128.into());
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_ir.rs:136]: Will remove, noting it here as it deserves a separate issue. It can only lead to bugs, if we create a register and then not store a value in it, or not store a value in a register whose index is larger than that register.

File: crates/noirc_evaluator/src/brillig/brillig_gen.rs [hunk 4]
```diff
@@ -199,20 +208,84 @@ impl BrilligGen {
...
+            1,
+            "array has a composite type as its element, which we do not support yet"
+        );
+        let atomic_type = element_type[0].clone();
+
+        match operator {
+            BinaryOp::Eq => {
+                // Allocate one array for the equality comparisons
+                let comparisons_array_ptr = self.context.create_register();
+                self.context.allocate_array(comparisons_array_ptr, num_elements, false);
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_gen.rs:264]: Seems we are doing something akin to inlining the bytecode since the num_elements is known -- we probably want to do it the way that assembly does it which is by using a for-loop with labels. I can push an unoptimized version

File: crates/noirc_evaluator/src/brillig/brillig_ir.rs [hunk 5]
```diff
@@ -43,14 +43,75 @@ impl BrilligContext {
...
+    pub(crate) fn allocate_array(
+        &mut self,
+        pointer_register: RegisterIndex,
+        size: u32,
+        prefilled: bool,
+    ) {
+        let allocation = self.memory.allocate(size as usize);
+
+        // If the array is prefilled (for example, parameter arrays), then we do not need to expand memory
+        if !prefilled {
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_ir.rs:55]: I can see how this fixes the bug, though I wonder if this is leaking an abstraction -- ie parameter arrays are not a concept in this part of the code.

Will investigate the ssa ir!

File: crates/noirc_evaluator/src/brillig/brillig_ir/memory.rs [hunk 6]
```diff
@@ -6,10 +7,34 @@ pub(crate) struct BrilligMemory {
...
 impl BrilligMemory {
-    pub(crate) fn allocate(&mut self, size: usize) -> usize {
+    /// Allocate an array of size `size`
+    pub(crate) fn allocate(&mut self, size: usize) -> Allocation {
         let start = self.free_mem_pointer;
         self.free_mem_pointer += size;
-        start
+        Allocation {
+            start_address: MemoryAddress(start),
+            end_address: MemoryAddress(self.free_mem_pointer - 1),
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/brillig/brillig_ir/memory.rs:37]: Good catch ser

File: crates/nargo_cli/tests/test_data_ssa_refactor/brillig_array_eq/src/main.nr [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Tests a very simple program.
+// 
+// The features being tested is using array equality on brillig
+fn main(x: [Field;3], y : [Field;3])  {
+    assert(array_eq(x, y));
+    assert(!array_eq(x, [1,2,4]));
```
Comment (kevaundray) [inline crates/nargo_cli/tests/test_data_ssa_refactor/brillig_array_eq/src/main.nr:6]: We probably want to add the edge case here, for empty arrays -- it should return true in that case.

An issue to open is to check if the ssa ir will give a compiler error on arrays of different lengths. If it gets to brillig ir, then it will panic in the `type_of_binary_operation` function

Comment (sirasistant) [reply crates/nargo_cli/tests/test_data_ssa_refactor/brillig_array_eq/src/main.nr:6]: Empty arrays don't seem to be allowed 
`error: Arrays must have at least one element`
And with different sizes the compiler issues
`error: Can only compare arrays of the same length. Here LHS is of length 3, and RHS is 2`

**Comment by jfecher** [comment]
I believe this PR will no longer be necessary when #1704 is merged

**Comment by kevaundray** [comment]
#1704 has been merged so closing
